### PR TITLE
bazel run :tests will succeed

### PR DESCRIPTION
**Pull request by erpang007chenfs**
## Proposed changes

Please include a description of the problem or feature this PR is addressing. If there is a corresponding issue, include the issue #.

## Checklist

Put an `x` in the boxes that apply.

- [ ] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [ ] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have updated the necessary documentation (if needed)

### CODE REVIEW

File: BUILD.bazel [hunk 1]
```diff
@@ -1,24 +1,38 @@
...
-load("//build_rules:kernel_preamble.bzl", "make_kernel_rules")
 
 package(default_visibility = ["//visibility:public"])
 
 # Generate preamble for common backend.
 genrule(
-    name = "generate_compiled_preamble_common",
-    srcs = glob(["mlx/backend/common/**/*"]),
+    name = "generate_preamble_common",
+    srcs = glob(["mlx/types/*"]) + glob(["mlx/backend/common/**/*"]),
```
Comment (xnorai) [inline BUILD.bazel:9]: Glob contains a list, you can specify both patterns in it.

File: BUILD.bazel [hunk 2]
```diff
@@ -63,6 +152,8 @@ cc_library(
             "mlx/backend/no_cpu/**/*",
             "mlx/backend/no_metal/**/*",
             "mlx/distributed/mpi/**/*",
+            "mlx/io/no_safetensors.cpp",  # TODO: Parameterize
```
Comment (xnorai) [inline BUILD.bazel:155]: I think these two should be omitted. Let's just always build with safe tensors and GGUF.

File: BUILD.bazel [hunk 3]
```diff
@@ -74,11 +165,12 @@ cc_library(
             "mlx/distributed/mpi/**/*",
         ],
     ),
+    data = [":kernels"],
     defines = [
         "ACCELERATE_NEW_LAPACK",
         "STRINGIFY_(x)=#x",
         "STRINGIFY(x)=STRINGIFY_(x)",
-        "METAL_PATH=STRINGIFY(mlx.metallib)",
+        "METAL_PATH=STRINGIFY(kernels.metallib)",  # TODO: Change library name to mlx.metallib
```
Comment (xnorai) [inline BUILD.bazel:173]: So how did it find it for the test?

File: build_rules/metal.bzl [hunk 4]
```diff
@@ -26,7 +26,7 @@ def _metal_library_impl(ctx):
             workspace_root = "."
 
         # Note that we allow includes wrt workspace root, as specified in Style.
-        args.add_all(["metal", "-c", src.path, "-o", air_file.path, "-I", workspace_root])
+        args.add_all(["metal", "-fno-fast-math", "-c", src.path, "-o", air_file.path, "-I", workspace_root])
```
Comment (xnorai) [inline build_rules/metal.bzl:29]: Is this also the setting in CMake? What's the default?

Comment (erpang007chenfs) [reply build_rules/metal.bzl]: xcrun -sdk macosx metal -Wall -Wextra -fno-fast-math -c /Users/chenfan_sun/workspace/mlx_2/mlx/backend/metal/kernels/conv.metal -I/Users/chenfan_sun/workspace/mlx_2 -o conv.air

File: BUILD.bazel (hunk 3 above)
Comment (erpang007chenfs) [reply BUILD.bazel:176]: We specify `data = [":kernels"],`, which is just `bazel-bin/kernels.metalib`.

File: BUILD.bazel (hunk 2 above)
Comment (erpang007chenfs) [reply BUILD.bazel]: Sure.

File: BUILD.bazel (hunk 1 above)
Comment (erpang007chenfs) [reply BUILD.bazel]: Done

**Review by erpang007chenfs — COMMENTED** [review]
PTAL. I'm going to address some of the TODOs

**Review by 1e100 — APPROVED** [review]

**Comment by 1e100** [comment]
LGTM, merging.

**Review by xnorai — APPROVED** [review]

**Pull request merged**